department-of-veterans-affairs/abd-vro
# Datetimes not being serialized/deserialized correctly

Contained in this PR:
1. Changes code to correctly serialize/deserialize fields expressed as a datetime, i.e., `2023-09-27T00:00:00-06:00`
2. This change forced refactor of putClaimLifecycleStatus status due to using the same Beans configured for the restTemplate/rabbitTemplate for message conversion.
  - Implements the putClaimLifecycleStatus endpoint as more of a passthrough implementation of the response received from the BIP API similar to the updates implemented in PR #2165 for the getClaimContentions endpoint.
4. Removal of endpoint to set the claim status to RFD, since the putClaimLifecycleStatus endpoint can handle that request


## What was the problem?

Date times are not being de/serialized into a date time expressed as a string, but instead as a number of seconds.

Associated tickets or Slack threads:
- #2091
- Stacked on PR:
  - #2237 


## How does this fix it?[^1]

Updates the jackson2JsonMessageConverter with an objectMapper configured to use the JavaTimeModule which supports serialization/deserialization of datetime strings into java Instants. 

## How to test this PR
- Run the unit tests via `./gradlew :svc-bip-api:check`
- Start the platform base docker containers to start RabbitMq and start the BIP Mock container (see [Docker Compose](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Docker-Compose))
  - run the integration tests via `./gradlew :svc-bip-api:integrationTest`
  - build and compose the `svc-bip-api` docker and run a test by:
     - create a new queue on the `bip-api` exchange called `response`
     - open to the rabbitmq console to the [putClaimLifecycleStatus](http://localhost:15672/#/queues/%2F/putClaimLifecycleStatus) 
     - publish a message with the body `{"claimId":1010,"claimLifecycleStatus":"test"}` and add a message property called `reply_to` with the value `response`
     - open the response queue page and perform the get message action
     - verify it is a 200 and the response is the mocked response
     - verify time fields
       - open to the rabbitmq console to the [getClaimDetailsQueue](http://localhost:15672/#/queues/%2F/getClaimDetailsQueue) 
       - publish a message with the body `{"claimId":1001}` and add a message property called `reply_to` with the value `response`
       - open the response queue page and perform the get message action
       - verify it is a 200 and the response is the mocked response
       - verify time fields are timestamp formats


[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).
